astacus/common/cassandra/schema.py

```diff
@@ -0,0 +1,336 @@
+
+        if metadata.final_func:
+            ret += "{}FINALFUNC {}".format(sep, cm.protect_name(metadata.final_func))
+        if metadata.initial_condition:
+            ret += f"{sep}INITCOND {metadata.initial_condition}"
+
+        return ret
+
+    @classmethod
+    def from_cassandra_metadata(cls, metadata: cm.Aggregate) -> "CassandraAggregate":
+        return cls(
+            name=metadata.name,
+            cql_create_self=cls.get_create_statement(metadata=metadata),
+            argument_types=metadata.argument_types,
+        )
+
+    def get_metadata_dict(self, keyspace_metadata: cm.KeyspaceMetadata):
+        return keyspace_metadata.aggregates
+
+
+class CassandraIndex(CassandraNamed):
+    # from_cassandra_metadata covered by superclass
+    pass
+
+
+class CassandraMaterializedView(CassandraNamed):
+    # from_cassandra_metadata covered by superclass
+    pass
+
+
+class CassandraTrigger(CassandraNamed):
+    # from_cassandra_metadata covered by superclass
+    pass
+
+
+class CassandraTable(CassandraNamed):
+    indexes: List[CassandraIndex]
+    materialized_views: List[CassandraMaterializedView]
+    triggers: List[CassandraTrigger]
+
+    @classmethod
+    def from_cassandra_metadata(cls, metadata: cm.TableMetadata) -> "CassandraTable":
+        return cls(
+            # CassandraNamed
+            name=metadata.name,
+            cql_create_self=metadata.as_cql_query(),
+            # CassandraTable
+            indexes=sorted(CassandraIndex.from_cassandra_metadata(metadata) for metadata in metadata.indexes.values()),
+            materialized_views=sorted(
+                CassandraMaterializedView.from_cassandra_metadata(metadata) for metadata in metadata.views.values()
+            ),
+            triggers=sorted(CassandraTrigger.from_cassandra_metadata(metadata) for metadata in metadata.triggers.values()),
+        )
+
+    def restore_post_data_in_keyspace(self, cas: CassandraSession, keyspace_metadata: cm.KeyspaceMetadata):
+        table_metadata = keyspace_metadata.tables[self.name]
+        for resource_map, resource_list in [
+            [table_metadata.indexes, self.indexes],
+            [table_metadata.triggers, self.triggers],
+            [table_metadata.views, self.materialized_views],
+        ]:
+            for resource in resource_list:
+                resource.restore_if_needed(cas, resource_map)
+
+
+class CassandraKeyspace(CassandraNamed):
+    # CQL to create everything
+    cql_create_all: str
+
+    aggregates: List[CassandraAggregate]
+    functions: List[CassandraFunction]
+    tables: List[CassandraTable]
+    user_types: List[CassandraUserType]
+
+    @classmethod
+    def from_cassandra_metadata(cls, metadata: cm.KeyspaceMetadata) -> "CassandraKeyspace":
+        return cls(
+            # CassandraNamed
+            name=metadata.name,
+            cql_create_self=metadata.as_cql_query(),
+            # CassandraKeyspace
+            cql_create_all=metadata.export_as_string(),
```

Comment:
It is mostly for debugging purposes - as API allows it, later on we can try to fix things if the recursive dump+restore we do actually has a bug in it. (Some text -> doesn't cost much to store, but may be useful in an emergency.)

Reply:
it's probably affected by BF-713: if a table has a default_ttl set on it and has a materialized view, this default_ttl setting leaks into CREATE statement for the MV causing it to fail. Cashew had a hack to circumvent that, but I don't see it here. I thought that since you used a different function in metadata to capture the MV definition maybe it's not an issue anymore, but worth checking, I guess.